### PR TITLE
added logging for equations in annotation generation

### LINKED ISSUE #229: Add logging for equations containing bold and italics
@HassanAkbar , I am concerned that if we remove the bold and italics in equations during translation, the maintainers of the standard will not be able to put them back in if they want to.

It is also the case that some current equations will work when translated to AsciiMath and MathML, and some will not; I think we would be adding value to ISO-10303 if we track them.

It also needs to be said that Metanorma follows the MathML default of italics for variables, and additional settings will be needed to default to plain instead of italics for variables, as seems to be implied in the current formulas. This is feasible as a setting, but it will not happen out of the box.

I propose the following course of action, @HassanAkbar ; @ronaldtse has already approved it (at least in principle):

1. Log all equations encountered containing boldface and italics, during translation of Express to Asciidoc, in a file that maintainers can access, with the following content:

* The raw Express formula, with boldface and italics tags
* The document it is in
* The nearest Express anchor, so we can find the equation in the converted text
* The Asciimath conversion of the Express formula, *including* bold and italics
* (Possibly) The Asciimath conversion of the Express formula, *without* bold and italics, which is what is actually going into the Asciidoc
* The status of whether that Asciimath formula including bold and italics is valid according to Plurimath: if it is not, that indicates that it needs to be fixed.

To do that validation, you will do what I am already doing in https://github.com/metanorma/metanorma-standoc/blob/f25f7d10f2c42c0772207a9104e02aa90a17648c/lib/metanorma/standoc/validate.rb#L34 ; I've changed the code to match what you will be doing.

````
     def math_validate(expr)
          Plurimath::Math.parse(expr, "asciimath").to_mathml
        rescue StandardError => e
          math_validate_error(math, m, e)
      end
````

So we take the asciimath, parse it into Plurimath's internal representation, and then generate MathML (as Metanorma eventually will). If something blows up in either stage, we want to know about it, because it points to a problem in the formula.

_Originally posted by @opoudjis in https://github.com/metanorma/stepmod-utils/issues/191#issuecomment-1817464353_

### PR DESCRIPTION
**Pull request by HassanAkbar**
Added logging for `equations` when generating annotations.

@opoudjis The output looks like this, Is this ok or do we need to change this?

```
I, [2023-12-20T15:55:12.464372 #12281]  INFO -- : 
=================== Equation Start ===================
Document: ../iso-10303-stepmod-wg12/data/resources/geometric_model_schema/descriptions.xml
Nearest Anchor: geometric_model_schema.manifold_solid_brep

Formula (original):
<eqn>
					<i>
						<b> (((B)[S])[F])[L] = (((B)[S])[F]){L}   </b>
					</i>
				</eqn>

------------------

Status: true
Formula (asciimath):

[stem]
++++
(((B)[S])[F])[L] = (((B)[S])[F]){L}
++++




------------------

Status: true
Formula (asciimath with bold and italics included):

[stem]
++++
ii(bb((((B)[S])[F])[L] = (((B)[S])[F]){L}))
++++




=================== Equation End ===================
```

closes #229

### CODE REVIEW

File: stepmod-utils.gemspec [hunk 1]
```diff
@@ -35,12 +35,14 @@ Gem::Specification.new do |spec|
...
   spec.add_runtime_dependency "glossarist", "~> 2.0"
   spec.add_runtime_dependency "indefinite_article"
+  spec.add_runtime_dependency "plurimath"
   spec.add_runtime_dependency "ptools"
   spec.add_runtime_dependency "pubid-iso"
   spec.add_runtime_dependency "octokit"
   spec.add_runtime_dependency "down"
   spec.add_runtime_dependency "reverse_adoc", ">= 0.3"
   spec.add_runtime_dependency "thor", ">= 0.20"
+  spec.add_runtime_dependency "unitsml"
```
Comment (ronaldtse) [inline stepmod-utils.gemspec:45]: I think Plurimath already includes `unitsml` so we don't need it here?

**Review by ronaldtse — APPROVED** [review]

File: stepmod-utils.gemspec (hunk 1 above)
Comment (HassanAkbar) [reply stepmod-utils.gemspec:45]: @ronaldtse Currently `unitsml` require `plurimath` so if we add `unitsml` as a dependency in `plurimath` it creates a circular dependency so it is not included in `plurimath` in the latest version.

**Review by opoudjis — CHANGES_REQUESTED** [review]
This is what I asked for, with two provisos:

1. If you haven't already done so, would be useful to have an ordinal number in the log for formulas, will help in citing errors: 1, 2, 3 etc. But this is not essential.
2. Does "status" mean whether the formula is valid or not? If so, don't use "true/false" but "valid/invalid". "True/false" is hard to interpret.

**Comment by HassanAkbar** [comment]
@opoudjis 
> If you haven't already done so, would be useful to have an ordinal number in the log for formulas, will help in citing errors: 1, 2, 3 etc. But this is not essential.

Sure, will add this.

> Does "status" mean whether the formula is valid or not? If so, don't use "true/false" but "valid/invalid". "True/false" is hard to interpret.

Yes `status` is calculated by passing the formula to `Plurimath`, I'll update it to `valid/invalid`

**Comment by HassanAkbar** [comment]
@opoudjis I've updated the PR to include your suggestions. The output now looks like below

```
=================== Equation 1 Start ===================
Document: ../iso-10303-stepmod-wg12/data/modules/tessellated_geometry/arm_descriptions.xml
Nearest Anchor: Tessellated_geometry_arm.Cubic_bezier_triangulated_face

Formula (original):
<eqn> <b>s</b>(u,v) = (1 - u -v)<sup>3</sup><b>P<sub>1</sub></b> + u<sup>3</sup><b>P<sub>2</sub></b> + v<sup>3</sup><b>P<sub>3</sub></b> + 3(1 - u -v)<sup>2</sup>u<b>P<sub>4</sub></b> + 3(1 - u -v)u<sup>2</sup><b>P<sub>5</sub></b> + 3u<sup>2</sup>v<b>P<sub>6</sub></b> + 3uv<sup>2</sup><b>P<sub>7</sub></b> + 3(1 - u -v)v<sup>2</sup><b>P<sub>8</sub></b> + 3(1 - u -v)<sup>2</sup>v<b>P<sub>9</sub></b> + 6uv(1 - u -v)<b>P<sub>10</sub></b> </eqn>

------------------

Status: valid
Formula (asciimath):

[stem]
++++
s(u,v) = (1 - u -v)^{3}P_{1} + u^{3}P_{2} + v^{3}P_{3} + 3(1 - u -v)^{2}uP_{4} + 3(1 - u -v)u^{2}P_{5} + 3u^{2}vP_{6} + 3uv^{2}P_{7} + 3(1 - u -v)v^{2}P_{8} + 3(1 - u -v)^{2}vP_{9} + 6uv(1 - u -v)P_{10}
++++




------------------

Status: valid
Formula (asciimath with bold and italics included):

[stem]
++++
bb(s)(u,v) = (1 - u -v)^{3}bb(P_{1}) + u^{3}bb(P_{2}) + v^{3}bb(P_{3}) + 3(1 - u -v)^{2}ubb(P_{4}) + 3(1 - u -v)u^{2}bb(P_{5}) + 3u^{2}vbb(P_{6}) + 3uv^{2}bb(P_{7}) + 3(1 - u -v)v^{2}bb(P_{8}) + 3(1 - u -v)^{2}vbb(P_{9}) + 6uv(1 - u -v)bb(P_{10})
++++




=================== Equation 1 End ===================
```

**Review by opoudjis — APPROVED** [review]

**Comment by HassanAkbar** [comment]
Released version `0.4.11` for this change.

**Pull request merged**